cmd/bump.go

```diff
@@ -0,0 +1,247 @@
+)
+
+type pullRequest struct {
+	Number int
+	Title  string
+	Kinds  changelog.Kinds
+}
+
+var (
+	labelPrefix = "changelog/"
+	skipLabel   = labelPrefix + "skip"
+)
+
+type changelogData struct {
+	Version      string
+	Date         string
+	PullRequests []pullRequest
+	Skipped      []pullRequest
+	Contributors []string
+}
+
+const changelogTmpl = `## {{ .Version }} / {{ .Date }}
+{{ range .PullRequests }}
+* [{{ .Kinds.String }}] {{ makeSentence .Title }} #{{ .Number }}
+{{- end }}
+<!-- Skipped pull requests:{{ range .Skipped }}
+* [{{ .Kinds.String }}] {{ makeSentence .Title }} #{{ .Number }}
+{{- end }} -->
+
+Contributors:
+{{ range .Contributors }}
+* @{{ . }}
+{{- end }}
+
+`
+
+func writeChangelog(w io.Writer, version string, prs, skippedPrs []pullRequest, contributors []string) error {
+	sort.SliceStable(prs, func(i int, j int) bool { return prs[i].Kinds.Before(prs[j].Kinds) })
+	sort.SliceStable(skippedPrs, func(i int, j int) bool { return skippedPrs[i].Kinds.Before(skippedPrs[j].Kinds) })
+	sort.Strings(contributors)
+
+	tmpl, err := template.New("changelog").Funcs(
+		template.FuncMap{
+			"makeSentence": func(s string) string {
+				s = strings.TrimRight(s, ".")
+				return s + "."
+			},
+		}).Parse(changelogTmpl)
+	if err != nil {
+		return errors.Wrap(err, "invalid template")
+	}
+
+	return tmpl.Execute(w, &changelogData{
+		Version:      version,
+		Date:         time.Now().Format("2006-01-02"),
+		PullRequests: prs,
+		Skipped:      skippedPrs,
+		Contributors: contributors,
+	})
+}
+
+func runBumpVersion(changelogPath, versionPath string, bumpLevel string, preRelease string, baseBranch string) error {
+	current, err := projInfo.ToSemver()
+	if err != nil {
+		return err
+	}
+
+	next := *current
+	switch bumpLevel {
+	case "major":
+		next = current.IncMajor()
+	case "minor":
+		next = current.IncMinor()
+	case "patch":
+		next = current.IncPatch()
+	}
+	next, err = next.SetPrerelease(preRelease)
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	if *timeout != time.Duration(0) {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, *timeout)
+		defer cancel()
+	}
+	client, err := githubUtil.NewClient(ctx)
+	if err != nil {
+		info(fmt.Sprintf("failed to create authenticated GitHub client: %v", err))
+		info("Fallback to client without unauthentication")
+		client = github.NewClient(nil)
+	}
+
+	lastTag := "v" + current.String()
+	commit, _, err := client.Repositories.GetCommit(ctx, projInfo.Owner, projInfo.Name, lastTag)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("Fail to get the GitHub commit for %s", lastTag))
+	}
+	lastTagTime := commit.GetCommit().GetCommitter().GetDate()
+	lastCommitSHA := commit.GetSHA()
+
+	// Gather all pull requests merged since the last tag.
+	var (
+		prs, skipped     []pullRequest
+		uniqContributors = make(map[string]struct{})
+	)
+	err = githubUtil.ReadAll(
+		func(opts *github.ListOptions) (*github.Response, error) {
```

Comment:
Is there a better way to do this? This is a pretty long and complicated anonymous func.